File: ir.py
from enum import Enum

import z3
import dsl
from functools import reduce


class Form(Enum):
    DNF = 0
    CNF = 1

class Ir:
    pass

class OrIr(Ir):
    def __init__(self, inner, toggle_var=None):
        self.inner = inner
        self.toggle_var = toggle_var

    def apply(self, env, io_example):
        return OrIr([node.apply(env, io_example) for node in self.inner], self.toggle_var)

    def to_z3(self, form):
        if form == Form.DNF:
            if self.toggle_var is None:
                return z3.Or([x.to_z3(form) for x in self.inner])
            else:
                return z3.Implies(self.toggle_var, z3.Or([x.to_z3(form) for x in self.inner]))
        elif form == Form.CNF:
            if self.toggle_var is None:
                return z3.Or(z3.Not(z3.Or([x.toggle_var for x in self.inner])), z3.Or([x.to_z3(form) for x in self.inner]))
            else:
                return z3.And(self.toggle_var, z3.Or([x.to_z3(form) for x in self.inner]))

    def toggle_var_sum(self):
        if self.toggle_var is not None:
            total = z3.If(self.toggle_var, 1, 0)
        else:
            total = 0
        for x in self.inner:
            inner_total = x.toggle_var_sum()
            if inner_total is not 0:
                total += inner_total
        return total

class AndIr(Ir):
    def __init__(self, inner, toggle_var=None):
        self.inner = inner
        self.toggle_var = toggle_var

    def apply(self, env, io_example):
        return AndIr([node.apply(env, io_example) for node in self.inner], self.toggle_var)

    def to_z3(self, form):
        if form == Form.DNF:
            if self.toggle_var is None:
                return z3.And(z3.Or([x.toggle_var for x in self.inner]), z3.And([x.to_z3(form) for x in self.inner]))
            else:
                return z3.Implies(self.toggle_var, z3.And([x.to_z3(form) for x in self.inner]))
        elif form == Form.CNF:
            if self.toggle_var is None:
                return z3.And([x.to_z3(form) for x in self.inner])
            else:
                return z3.And(self.toggle_var, z3.And([x.to_z3(form) for x in self.inner]))

    def toggle_var_sum(self):
        if self.toggle_var is not None:
            total = z3.If(self.toggle_var, 1, 0)
        else:
            total = 0
        for x in self.inner:
            inner_total = x.toggle_var_sum()
            if inner_total is not 0:
                total += inner_total
        return total

class AnyIr(Ir):
    def __init__(self, object_var, inner, toggle_var):
        self.object_var = object_var
        self.inner = inner
        self.toggle_var = toggle_var

    def apply(self, env, io_example):
        return OrIr([self.inner.apply(env | {self.object_var: io_example.get_base(box)}, io_example) for box in io_example], self.toggle_var)

    def to_z3(self, form):
        raise Exception("Unable to convert partially compiled formula to z3 form. This formula contains an any expression. Try compiling to completely reify all any statements")

    def toggle_var_sum(self):
        if self.toggle_var is not None:
            return z3.If(self.toggle_var, 1, 0) + self.inner.toggle_var_sum()
        else:
            return self.inner.toggle_var_sum()

class AllIr(Ir):
    def __init__(self, object_var, inner, toggle_var):
        self.object_var = object_var
        self.inner = inner
        self.toggle_var = toggle_var

    def apply(self, env, io_example):
        return AndIr([self.inner.apply(env | {self.object_var: io_example.get_base(box)}, io_example) for box in io_example], self.toggle_var)

    def to_z3(self, form):
        raise Exception("Unable to convert partially compiled formula to z3 form. This formula contains an all expression. Try compiling to completely reify all all statements")

    def toggle_var_sum(self):
        if self.toggle_var is not None:
            return z3.If(self.toggle_var, 1, 0) + self.inner.toggle_var_sum()
        else:
            return self.inner.toggle_var_sum()

var_index = 0

def get_fresh_toggle_var():
    global var_index
    ret = z3.Bool("v" + str(var_index))
    var_index += 1
    return ret

class BooleanIr(Ir):
    def __init__(self, value, toggle_var):
        self.value = value
        self.toggle_var = toggle_var

    def apply(self, env, io_example):
        return self

    def to_z3(self, form):
        if form == Form.DNF:
            return z3.Implies(self.toggle_var, z3.Bool(self.value))
        elif form == Form.CNF:
            return z3.And(self.toggle_var, z3.Bool(self.value))

    def toggle_var_sum(self):
        if self.toggle_var is not None:
            return z3.If(self.toggle_var, 1, 0)
        else:
            return 0

class MatchIr(Ir):
    def __init__(self, object_a, object_b, negated, toggle_var):
        self.object_a = object_a
        self.object_b = object_b
        self.toggle_var = toggle_var
        self.negated = negated

    def apply(self, env, io_example):
        if self.object_a in env:
            a = env[self.object_a]
        else:
            a = self.object_a
        if self.object_b in env:
            b = env[self.object_b]
        else:
            b = self.object_b
        if isinstance(a, dsl.ObjectLiteral) and isinstance(b, dsl.ObjectLiteral):
            if self.negated:
                return BooleanIr(a != b, self.toggle_var)
            else:
                return BooleanIr(a == b, self.toggle_var)
        else:
            return MatchIr(a, b, self.negated, self.toggle_var)

    def to_z3(self, toggle_var_constructor):
        raise Exception("Unable to convert partially compiled formula to z3 form. This formula contains a match expression. Try compiling to completely reify all match statements")

    def toggle_var_sum(self):
        if self.toggle_var is not None:
            return z3.If(self.toggle_var, 1, 0)
        else:
            return 0

def compile(ir, z3_solution):
    if isinstance(ir, MatchIr):
        if ir.toggle_var is None or bool(z3_solution[ir.toggle_var]):
            m = dsl.Match(ir.object_a, ir.object_b)
            if ir.negated:
                return dsl.NotBool(m)
            else:
                return m
        else:
            return None
    elif isinstance(ir, AnyIr):
        if ir.toggle_var is None or bool(z3_solution[ir.toggle_var]):
            return dsl.Any(dsl.Predicate(ir.object_var, compile(ir.inner, z3_solution)))
        else:
            return None
    elif isinstance(ir, AllIr):
        if ir.toggle_var is None or bool(z3_solution[ir.toggle_var]):
            return dsl.All(dsl.Predicate(ir.object_var, compile(ir.inner, z3_solution)))
        else:
            return None
    elif isinstance(ir, OrIr):
        if ir.toggle_var is None or bool(z3_solution[ir.toggle_var]):
            compiled = [compile(x, z3_solution) for x in ir.inner]
            compiled = [x for x in compiled if x is not None]
            if len(compiled) > 0:
                return reduce(dsl.OrBool, compiled)
            else:
                return None
        else:
            return None
    elif isinstance(ir, AndIr):
        if ir.toggle_var is None or bool(z3_solution[ir.toggle_var]):
            compiled = [compile(x, z3_solution) for x in ir.inner]
            compiled = [x for x in compiled if x is not None]
            if len(compiled) > 0:
                return reduce(dsl.AndBool, compiled)
            else:
                return None
        else:
            return None

File: DataLabeling/Ir.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Z3;
using System.Collections.Immutable;

namespace DataLabeling
{
    public enum Form { DNF, CNF }

    public abstract class Ir
    {
        public BoolExpr ToggleVar;

        protected Ir(BoolExpr toggleVar) {
            ToggleVar = toggleVar;
        }

        public abstract Ir Apply(ImmutableDictionary<ObjectVariable, Tuple<BoundingBox, ObjectLiteral>> env, IOExample example);

        public abstract BoolExpr ToZ3(Context ctx, Form form);

        public abstract ArithExpr? ToggleVarSum(Context ctx);

        public abstract BooleanAst? Compile(Model z3Solution);

        protected bool ToggleSolution(Model z3Solution) {
            return z3Solution.Eval(ToggleVar).BoolValue == Z3_lbool.Z3_L_TRUE;
        }
    }

    public class OrIr : Ir
    {
        public List<Ir> Inner { get; private set; }

        public OrIr(List<Ir> inner, BoolExpr toggleVar) : base(toggleVar) {
            Inner = inner;
        }

        public override Ir Apply(ImmutableDictionary<ObjectVariable, Tuple<BoundingBox, ObjectLiteral>> env, IOExample example) {
            return new OrIr(Inner.Select(node => node.Apply(env, example)).ToList(), ToggleVar);
        }

        public override ArithExpr ToggleVarSum(Context ctx) {
            if (ToggleVar != null) {
                ArithExpr total = (ArithExpr)ctx.MkITE(ToggleVar, ctx.MkInt(1), ctx.MkInt(0));
                foreach (Ir r in Inner) {
                    total = total + r.ToggleVarSum(ctx);
                }
                return total;
            } else {
                ArithExpr? total = null;
                foreach (Ir r in Inner) {
                    ArithExpr? res = r.ToggleVarSum(ctx);
                    if (total == null) {
                        total = res;
                    } else if (res != null) {
                        total = total + r.ToggleVarSum(ctx);
                    }
                }
                return total;
            }
        }

        public override BoolExpr ToZ3(Context ctx, Form form) {
            if (form == Form.DNF) {
                if (ToggleVar == null) {
                    // This Or clause is in the outermost form, so we should ignore any toggle variable behaviour
                    return ctx.MkOr(Inner.Select(x => x.ToZ3(ctx, form)));
                } else {
                    // This Or clause is embedded within an And clause as part of an "Any"
                    return ctx.MkImplies(ToggleVar, ctx.MkOr(Inner.Select(x => x.ToZ3(ctx, form))));
                }
            } else if (form == Form.CNF) {
                if (ToggleVar == null) {
                    // This Or clause is embedded within an and clause, so we should compute an equivalent toggle variable
                    return ctx.MkOr(ctx.MkNot(ctx.MkOr(Inner.Select(x => x.ToggleVar))), ctx.MkOr(Inner.Select(x => x.ToZ3(ctx, form))));
                } else {
                    // This Or clause is embedded within an Or clause as part of an "Any"
                    return ctx.MkAnd(ToggleVar, ctx.MkOr(Inner.Select(x => x.ToZ3(ctx, form))));
                }
            }

            throw new NotImplementedException();
        }

        public override BooleanAst? Compile(Model z3Solution) {
            if (ToggleVar == null || ToggleSolution(z3Solution)) {
                List<BooleanAst> compiled = Inner.Select(x => x.Compile(z3Solution)).Where(x => x != null).ToList();
                if (compiled.Count > 0) {
                    return compiled.Aggregate((BooleanAst left, BooleanAst right) => new OrBool(left, right));
                } else {
                    return null;
                }
            } else {
                return null;
            }
        }
    }

    public class AndIr : Ir
    {
        public List<Ir> Inner { get; private set; }

        public AndIr(List<Ir> inner, BoolExpr toggleVar) : base(toggleVar) {
            Inner = inner;
        }

        public override Ir Apply(ImmutableDictionary<ObjectVariable, Tuple<BoundingBox, ObjectLiteral>> env, IOExample example) {
            return new AndIr(Inner.Select(node => node.Apply(env, example)).ToList(), ToggleVar);
        }

        public override ArithExpr? ToggleVarSum(Context ctx) {
            if (ToggleVar != null) {
                ArithExpr total = (ArithExpr)ctx.MkITE(ToggleVar, ctx.MkInt(1), ctx.MkInt(0));
                foreach (Ir r in Inner) {
                    total = total + r.ToggleVarSum(ctx);
                }
                return total;
            } else {
                ArithExpr? total = null;
                foreach (Ir r in Inner) {
                    ArithExpr? res = r.ToggleVarSum(ctx);
                    if (total == null) {
                        total = res;
                    } else if (res != null) {
                        total = total + r.ToggleVarSum(ctx);
                    }
                }
                return total;
            }
        }

        public override BoolExpr ToZ3(Context ctx, Form form) {
            if (form == Form.DNF) {
                if (ToggleVar == null) {
                    // This And clause is embedded within an Or clause, so we should compute the equivalent toggle variable
                    return ctx.MkAnd(ctx.MkOr(Inner.Select(x => x.ToggleVar)), ctx.MkAnd(Inner.Select(x => x.ToZ3(ctx, form))));
                } else {
                    // This And clause is embedded within an And clause as part of an "All"
                    return ctx.MkImplies(ToggleVar, ctx.MkAnd(Inner.Select(x => x.ToZ3(ctx, form))));
                }
            } else if (form == Form.CNF) {
                if (ToggleVar == null) {
                    // This And clause in the outermost form, and we should ignore any toggles
                    return ctx.MkAnd(Inner.Select(x => x.ToZ3(ctx, form)));
                } else {
                    // This And clause is embedded within an Or clause as part of an "All"
                    return ctx.MkAnd(ToggleVar, ctx.MkAnd(Inner.Select(x => x.ToZ3(ctx, form))));
                }
            }

            throw new NotImplementedException();
        }

        public override BooleanAst? Compile(Model z3Solution) {
            if (ToggleVar == null || ToggleSolution(z3Solution)) {
                List<BooleanAst> compiled = Inner.Select(x => x.Compile(z3Solution)).Where(x => x != null).ToList();
                if (compiled.Count > 0) {
                    return compiled.Aggregate((BooleanAst left, BooleanAst right) => new AndBool(left, right));
                } else {
                    return null;
                }
            } else {
                return null;
            }
        }
    }

    public class AnyIr : Ir
    {
        public readonly ObjectVariable ObjVar;
        public readonly Ir Inner;

        public AnyIr(ObjectVariable objVar, Ir inner, BoolExpr toggleVar) : base(toggleVar) {
            ObjVar = objVar;
            Inner = inner;
        }

        public override Ir Apply(ImmutableDictionary<ObjectVariable, Tuple<BoundingBox, ObjectLiteral>> env, IOExample example) {
            return new OrIr(example.GetBoxes().Select(box => Inner.Apply(env.Add(ObjVar, Tuple.Create(box, example.GetBase(box))), example)).ToList(), ToggleVar);
        }

        public override ArithExpr? ToggleVarSum(Context ctx) {
            if (ToggleVar != null) {
                return (ArithExpr)ctx.MkITE(ToggleVar, ctx.MkInt(1), ctx.MkInt(0)) + Inner.ToggleVarSum(ctx);
            } else {
                return Inner.ToggleVarSum(ctx);
            }
        }

        public override BoolExpr ToZ3(Context ctx, Form form) {
            throw new NotImplementedException("Unable to convert partially compiled formula to z3 form. This formula contains an any expression. Try compiling to completely reify all any statements");
        }

        public override BooleanAst? Compile(Model z3Solution) {
            if (ToggleVar == null || ToggleSolution(z3Solution)) {
                return new Any(new PredicateLambda(ObjVar, Inner.Compile(z3Solution)));
            } else {
                return null;
            }
        }
    }

    public class AllIr : Ir
    {
        public readonly ObjectVariable ObjVar;
        public readonly Ir Inner;

        public AllIr(ObjectVariable objVar, Ir inner, BoolExpr toggleVar) : base(toggleVar) {
            ObjVar = objVar;
            Inner = inner;
        }

        public override Ir Apply(ImmutableDictionary<ObjectVariable, Tuple<BoundingBox, ObjectLiteral>> env, IOExample example) {
            return new AndIr(example.GetBoxes().Select(box => Inner.Apply(env.Add(ObjVar, Tuple.Create(box, example.GetBase(box))), example)).ToList(), ToggleVar);
        }

        public override ArithExpr? ToggleVarSum(Context ctx) {
            if (ToggleVar != null) {
                return (ArithExpr)ctx.MkITE(ToggleVar, ctx.MkInt(1), ctx.MkInt(0)) + Inner.ToggleVarSum(ctx);
            } else {
                return Inner.ToggleVarSum(ctx);
            }
        }

        public override BoolExpr ToZ3(Context ctx, Form form) {
            throw new NotImplementedException("Unable to convert partially compiled formula to z3 form. This formula contains an all expression. Try compiling to completely reify all all statements");
        }

        public override BooleanAst? Compile(Model z3Solution) {
            if (ToggleVar == null || ToggleSolution(z3Solution)) {
                return new All(new PredicateLambda(ObjVar, Inner.Compile(z3Solution)));
            } else {
                return null;
            }
        }
    }

    public class BooleanIr : Ir
    {
        public readonly bool Value;

        public BooleanIr(bool value, BoolExpr toggleVar) : base(toggleVar) {
            Value = value;
        }

        public override Ir Apply(ImmutableDictionary<ObjectVariable, Tuple<BoundingBox, ObjectLiteral>> env, IOExample example) {
            return this;
        }

        public override ArithExpr? ToggleVarSum(Context ctx) {
            if (ToggleVar != null) {
                return (ArithExpr) ctx.MkITE(ToggleVar, ctx.MkInt(1), ctx.MkInt(0));
            } else {
                return null;
            }
        }

        public override BoolExpr ToZ3(Context ctx, Form form) {
            if (form == Form.DNF) {
                return ctx.MkImplies(ToggleVar, ctx.MkBool(Value));
            } else {
                return ctx.MkAnd(ToggleVar, ctx.MkBool(Value));
            }
        }

        public override BooleanAst? Compile(Model z3Solution) {
            throw new NotImplementedException("Cannot compile a boolean to an AST representation");
        }
    }

    public class MatchIr : Ir
    {
        public readonly ObjectAst ObjA;
        public readonly ObjectAst ObjB;
        public readonly bool Negated;

        public MatchIr(ObjectAst objA, ObjectAst objB, bool negated, BoolExpr toggleVar) : base(toggleVar) {
            ObjA = objA;
            ObjB = objB;
            Negated = negated;
        }

        public override Ir Apply(ImmutableDictionary<ObjectVariable, Tuple<BoundingBox, ObjectLiteral>> env, IOExample example) {
            ObjectAst a = ObjA;
            ObjectAst b = ObjB;

            if ((ObjA is ObjectVariable) && env.ContainsKey((ObjectVariable) ObjA)) {
                a = env[(ObjectVariable)ObjA].Item2;
            }
            if ((ObjB is ObjectVariable) && env.ContainsKey((ObjectVariable) ObjB)) {
                b = env[(ObjectVariable)ObjB].Item2;
            }

            if ((a is ObjectLiteral) && (b is ObjectLiteral)) {
                bool eq = ((ObjectLiteral)a).Equals((ObjectLiteral)b);
                if (Negated) {
                    return new BooleanIr(!eq, ToggleVar);
                } else {
                    return new BooleanIr(eq, ToggleVar);
                }
            } else {
                return new MatchIr(a, b, Negated, ToggleVar);
            }
        }

        public override ArithExpr? ToggleVarSum(Context ctx) {
            if (ToggleVar != null) {
                return (ArithExpr)ctx.MkITE(ToggleVar, ctx.MkInt(1), ctx.MkInt(0));
            } else {
                return null;
            }
        }

        public override BoolExpr ToZ3(Context ctx, Form form) {
            throw new NotImplementedException("Unable to convert partially compiled formula to z3 form. This formula contains a match expression. Try compiling to completely reify all match statements");
        }

        public override BooleanAst? Compile(Model z3Solution) {
            if (ToggleVar == null || ToggleSolution(z3Solution)) {
                Match m = new Match(ObjA, ObjB);
                if (Negated) {
                    return new NotBool(m);
                } else {
                    return m;
                }
            } else {
                return null;
            }
        }
    }

    public class GeqIr : Ir
    {
        public readonly double Val;
        public readonly RealExpr ComparisonVar;

        public GeqIr(double val, RealExpr comparisonVar, BoolExpr toggleVar) : base(toggleVar) {
            Val = val;
            ComparisonVar = comparisonVar;
        }

        public override Ir Apply(ImmutableDictionary<ObjectVariable, Tuple<BoundingBox, ObjectLiteral>> env, IOExample example) {
            return this;
        }

        public override ArithExpr? ToggleVarSum(Context ctx) {
            if (ToggleVar != null) {
                return (ArithExpr)ctx.MkITE(ToggleVar, ctx.MkInt(1), ctx.MkInt(0));
            } else {
                return null;
            }
        }

        public override BoolExpr ToZ3(Context ctx, Form form) {
            if (form == Form.DNF) {
                return ctx.MkImplies(ToggleVar, ctx.MkReal(Val.ToString()) >= ComparisonVar);
            } else if (form == Form.CNF) {
                return ctx.MkAnd(ToggleVar, ctx.MkReal(Val.ToString()) >= ComparisonVar);
            }

            throw new NotImplementedException();
        }

        public override BooleanAst? Compile(Model z3Solution) {
            throw new NotImplementedException("Cannot compile a boolean to an AST representation");
        }
    }

    public class IOUIr : Ir
    {
        public readonly ObjectVariable ObjA;
        public readonly ObjectVariable ObjB;
        public RealExpr ComparisonVar;

        public IOUIr(ObjectVariable objA, ObjectVariable objB, RealExpr comparisonVar, BoolExpr toggleVar) : base(toggleVar) {
            ObjA = objA;
            ObjB = objB;
            ComparisonVar = comparisonVar;
        }

        public override Ir Apply(ImmutableDictionary<ObjectVariable, Tuple<BoundingBox, ObjectLiteral>> env, IOExample example) {
            if (env.ContainsKey(ObjA) && env.ContainsKey(ObjB)) {
                BoundingBox boxA = env[ObjA].Item1;
                BoundingBox boxB = env[ObjB].Item1;
                return new GeqIr(boxA.JaccardIndex(boxB), ComparisonVar, ToggleVar);
            } else {
                throw new KeyNotFoundException("Unable to find keys in environment when applying an IOU node");
            }
        }

        public override ArithExpr? ToggleVarSum(Context ctx) {
            if (ToggleVar != null) {
                return (ArithExpr) ctx.MkITE(ToggleVar, ctx.MkInt(1), ctx.MkInt(0));
            } else {
                return null;
            }
        }

        public override BoolExpr ToZ3(Context ctx, Form form) {
            throw new NotImplementedException("Unable to convert partially compiled formula to z3 form. This formula contains a IOU expression. Try compiling to completely reify all IOU statements");
        }

        public override BooleanAst? Compile(Model z3Solution) {
            if (ToggleVar == null || ToggleSolution(z3Solution)) {
                return new IOU(ObjA, ObjB, ((RatNum)z3Solution.Eval(ComparisonVar)).Double);
            } else {
                return null;
            }
        }
    }
}
